crates/matrix-sdk-common/src/js_tracing.rs

```diff
@@ -13,86 +13,207 @@
 
 //! Utilities for `tracing` in JS environments
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    fmt,
+    fmt::Debug,
+    io,
+    rc::Rc,
+    sync::atomic::{AtomicU32, Ordering},
+};
 
-use std::{fmt, fmt::Write};
-
-use tracing::{
-    field::{Field, Visit},
-    Event, Level, Subscriber,
+use tracing::{level_filters::LevelFilter, Level, Metadata};
+use tracing_subscriber::fmt::{
+    format::{DefaultFields, Format, Pretty},
+    MakeWriter, Subscriber,
 };
-use tracing_subscriber::layer::Context;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+extern "C" {
+    /// The type of a javascript-side `Logger` object which we can use to log
+    /// from the rust side.
+    pub type JsLogger;
+
+    #[wasm_bindgen(method)]
+    pub fn debug(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn info(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn warn(this: &JsLogger, data: &JsValue);
+
+    #[wasm_bindgen(method)]
+    pub fn error(this: &JsLogger, data: &JsValue);
+}
 
-/// An implementation of `tracing_subscriber::layer::Layer` which directs all
-/// events to the JS console
-#[derive(Debug, Default)]
-pub struct Layer {}
+/// A [`MakeWriter`] which will construct an [`io::Write`] instance that will
+/// write to either a [`JsLogger`] or the JavaScript console.
+///
+/// You can either construct this as part of an entire [`Subscriber`], via
+/// [`make_tracing_subscriber`], but it's also possible to
+/// call [`MakeJsLogWriter::new`] or [`MakeJsLogWriter::new_with_logger`] and
+/// then feed the result into
+/// [`tracing_subscriber::fmt::SubscriberBuilder::with_writer`], for example:
+///
+/// ```
+/// # use matrix_sdk_common::js_tracing::MakeJsLogWriter;
+/// # use tracing_subscriber::util::SubscriberInitExt;
+/// let subscriber =
+///     tracing_subscriber::fmt().with_writer(MakeJsLogWriter::new()).finish();
+/// subscriber.init();
+/// ```
+#[derive(Debug)]
+pub struct MakeJsLogWriter {
+    /// The logger to send messages to, or `None` for the console.
+    ///
+    /// Since [`MakeWriter`]s need to be [`Send`], and references to JS-side
+    /// objects are not `Send`, we cannot refer directly to the logger here. In
+    /// practice, the lack of `Send` shouldn't be a problem, because there
+    /// will only ever be one thread in the WASM environment, but the types
+    /// don't know that.
+    ///
+    /// So, we indirect through a thread-local hashmap instance. Each time we
+    /// construct a new `MakeJsLogWriter` backed by a `JsLogger`, we assign
+    /// it a unique ID; we then store that ID in the thread-local hashmap.
+    /// Then, when we come to use the logger, provided we are really
+    /// in the same thread, we can look up the logger in the map.
+    logger_id: Option<u32>,
+}
+
+thread_local!(static LOGGER_MAP: RefCell<HashMap<u32, Rc<JsLogger>>> = RefCell::new(HashMap::new()));
+static NEXT_LOGGER_ID: AtomicU32 = AtomicU32::new(0);
```

Comment:
I find it pretty unusual for `thread_local` to be used with (expression-like) round parentheses. Also, shouldn't the `NEXT_LOGGER_ID` also be thread-local for consistency?

```suggestion
thread_local! {
    static LOGGER_MAP: RefCell<HashMap<u32, Rc<JsLogger>>> = RefCell::new(HashMap::new());
    static NEXT_LOGGER_ID: Cell<u32> = Cell::new(0);
}
```

Reply:
> I find it pretty unusual for `thread_local` to be used with (expression-like) round parentheses.

Fair enough. I don't know where I cargo-culted the parentheses from but will get rid of them.

> Also, shouldn't the `NEXT_LOGGER_ID` also be thread-local for consistency?

Mmm, no, I think it's better it is not. If we make it thread-local then that means that, if we do somehow end up running this in a multithreaded environment, then different `MakeJsLogWriter` instances could end up using the same `id`. Then, if those instances get used from the "wrong" thread, they could silently write to the wrong `JsLogger`.

As things stand, with a globally-unique ID, we can detect that fault early and complain about it.

Reply:
Okay, makes sense 👍🏼